### DISCUSSION TITLE
how to use watcher with WatchEvent

### DISCUSSION THREAD

**Discussion by schmurfy** [kube-rs/kube#1253] [General]
Hi,
I am learning both rust and kube-rs at the same time so I admit some things are still confusing for me despite reading most of the rust book once.

I am trying to monitor events related to multiple kube entities, I first started with [API.watch](https://docs.rs/kube/latest/kube/struct.Api.html#method.watch) which led me to [watcher](https://docs.rs/kube-runtime/0.84.0/kube_runtime/watcher/fn.watcher.html) and that's where I am but...

the documentation for API.watch says that watcher is a most robust way to do that and the example here show nicely how to batch multiple streams (pods, deployments in my case) but they don't return the same objects, when using API.watch I get added, deleted, updated but when using the watcher I get only the resulting objects.

is there a way to do https://github.com/kube-rs/kube/blob/main/examples/multi_watcher.rs but keeping the added/updated/deleted information ?

Thanks for any pointer in the right direction :)

**Comment by schmurfy** [comment]
After digging more into this I found out that `applied_objects` is the culprit for dropping that information only keeping Added/Updated items, I saw `touched_objects` and playing with it but I failed to produce a working code without using any of these two to keep the full Event.

**Comment by clux** [comment]
Hey there, we do indeed discard the root event type as you've noticed for watcher streams.

This is because it's a bit of an anti-pattern to try to build specific behaviour against the enum variants because Kubernetes does not guarantee delivery of events (i.e. when your app is down, or have network desyncs) and this is a pretty common scenario. In particular;

- missing delete events means you'll never get a chance to do your cleanup logic
- added and modified is effectively same event if you missed the modify the first time

Thus, it is better to either say you want all Added + Updated events (through `applied_objects()`) when you want to check stuff related to active objects (sounds like your use-case), or ask for all event types (through `.touched_objects()`) and build idempotent controllers around those streams. If you need to do cleanup on `WatchEvent::Deleted` you should use finalizers ([example](https://github.com/kube-rs/kube/blob/main/examples/secret_syncer.rs)).

**Reply by schmurfy** [reply]
Thanks for the explanation, I am not writing an in cluster application but a cluster state viewer and wanted to keep an internal state of the cluster but I guess my choices are to either have a separate watcher for each type or do a periodic refresh.
I saw the reflector api which is essentially what I wanna achieve but I am using tauri and I want the data on the js side so I don't think it can help me much.

**Reply by clux** [reply]
hmm, yeah. I can actually see that in the cases where you are just tailing information for a UI and want super-up-to-date information, then it probably makes sense to flatten the stream without truncating the event information. Maybe this is something we should allow after all. It is possible to create something akin to [`EventFlatten`](https://docs.rs/kube-runtime/0.84.0/src/kube_runtime/utils/event_flatten.rs.html) (that is used by .applied_objects or .touched_objects) that flattens but maintains the raw `WatchEvent` objects provided we caveat such a helper.

as a stop-gap, if tauri does not constrain you, something like a dumb [`Store::state`](https://docs.rs/kube/latest/kube/runtime/reflector/struct.Store.html#method.state) dump with a periodic refresh would probably work well enough (just wouldn't be very responsive unless you cranked the loop frequency of that sync job up).

**Reply by schmurfy** [reply]
I will just use multiple API.watch for now and will see later if I can/want to improve it.
